README.md
## Gopei2 (Go Programming Environment Installer)

[![version](https://img.shields.io/badge/version-2.9.0-blue.svg)](https://github.com/geosoft1/tools/archive/master.zip)
[![license](https://img.shields.io/badge/license-gpl-blue.svg)](https://github.com/geosoft1/tools/blob/master/LICENSE)
![go](https://img.shields.io/badge/go-1.18.2-green.svg)
![liteide](https://img.shields.io/badge/liteide-38.0-orange.svg)

Gopei shell install [Go](http://golang.org) compiler, [LiteIDE](https://github.com/visualfc/liteide) and configure for you the entire environment, variables, paths, ide interface and even link the machine with your Github account in a few minutes with just one command.

You don't need to use complicated commands on Mac or Linux so you just have to write go programs. 

Also, add some high level commands wich help you easy make a repository, clone, push or get a project and more.

This project is third party. For support see [golang.org](http://golang.org) and [github.com/visualfc/liteide](https://github.com/visualfc/liteide).

Read the user manual on the [wiki page](https://github.com/geosoft1/tools/wiki). For cloud tool read [here](https://github.com/geosoft1/tools/wiki/Cloud-tool).

**Features**

- Install a stable version of Go compiler and LiteIDE on almost any Unix/Linux based desktops and servers.
- Create a simple and practical layout for ide (liteide.ini.mini).
- Also create `GOPATH` and add a hello world project.
- Add a nice launcher to dock or desktop.
- Extend snippets, project templates and themes.
- Add some useful cloud scripts for
   - working with github (be sure you have installed `git` before and a valid acount)
   - servers in cloud (see [Quantum Teleporter](https://github.com/geosoft1/tools/wiki/Cloud-tool)).
- Gopei shell is free as in free speech and free beer and always will be.

**How to use**

[Download](https://github.com/geosoft1/tools/archive/master.zip) anywhere and unarchive. From Terminal go to the `tools-master` folder and simply run:

    ./gopei

or if you have a github account

    .gopei -g

Use `` -h `` switch to see all options.

Note that your system must have `bash`,`curl` and `git` installed before. Also, some systems may need additional [packages](https://github.com/geosoft1/tools/wiki#platform-specific-information).


CHANGES
What's new in 1.6 The Convergence (milestone)

- complete code rewritten, new approach.
- versioning acording to [semantic versioning](http://semver.org), see `dfc` [manifest](http://dave.cheney.net/2016/06/24/gophers-please-tag-your-releases).
- go environment moved into `.go_profile` file also inspired from `dfc` [~/.bashrc](https://forum.golangbridge.org/t/where-to-place-tools/2430/5)

1.6.1
- some minor bugs corrections

What's new in 1.7 The Convergence (milestone)

- add unified github tool
- add cloud tool (Quantum Teleporter)

1.7.1
- switch to LiteIDE X30 because of incompatibilities between go1.7 and gocode in X29
- add some conditional compilation commands on ctrl+`

1.7.2
- old monaco.ttf url no more exist so switch to other url
- add armv7 support

1.7.3
- cloud tool add port forwarding in ssh connections

1.7.4
- cloud tool add detach command
- add chmod 700 .ssh for connections to OS X

1.7.5
- add nightly build tool
- godoc command in ide on ctrl+`

1.7.6
- add backup tool

1.8.0
- switch to LiteIDE X30.3 because of incompatibilities between go1.8 and gocode in X30
- corrects some issues with LiteIDE path in OS X

1.9.0
- switch to LiteIDE X32.2 because of incompatibilities between go1.9 and gocode

1.9.1
- corrects some issues with Ubuntu 17+

1.9.2
- cloud tool 3.3.0 add delete command

1.9.3
- resolve some changes on golang.org site
- unity/gnome (U18.04) launcher corrections

What's new in Gopei2 (milestone)

- mostly rewritten, remove obsolete tools/install and Monaco font
- add go get golang.org/x/tools/cmd/present in ctrl+` list
- update links to new LiteIDE versions based on Qt5 (witch to LiteIDE 34.2)
- improve aspect on Linux and Mac, new liteide.ini.mini
- extend project templates
- include new polished themes and color schemes
- extend snippets (eg. web handler )
- add cloud4 tool

2.1.0
- update to liteide35.5

2.2.0
- fixed versions go1.11.5 liteide35.5

2.3.0
- update to go1.12

2.3.1
- add snippets for tags (eg. `json:""`, `xml:""`)

2.4.0
- update to go1.12.5 liteide36

2.5.0
- update to go1.13

2.5.1
- update git commands with git reset, remove obsolete godoc

2.6.0
- update copyright for all tools
- GITTOKEN replace GITPASSWORD due github security policy changes
- small code rewiew for all tools

2.7.0
- update to go1.15 liteide37.1

2.8.0
- update to go1.16 liteide37.4
- add missing gopher icons in go/doc

2.9.0
- update to go1.18.2 and liteide38.0


Install2.sh
#!/bin/bash
#
# golang programming environment installer
# NOTE:
# this project is third party. for support see http://golang.org and http://code.google.com/p/golangide/
# author       : geosoft1@gmail.com
# license      : GPLv3
# version      : 2.0.0.1
#
# 07.08.2014   - start project, skeleton, tests (1.0.0.1).
# 08.08.2014   - added $DESKTOP_SESSION,kernel name and 64bit support (1.0.0.2),eye candy (1.0.0.3).
# 10.08.2014   - better integration with unity launcher bar. compact generated files (1.0.0.4).
# 17.08.2014   - experimental git support (2.0.0.1).
#
# TODO:
# +some Unity2d improvements
# +shortcut for other desktop environments
# -adding zenity for more interactivity with the user
# wget http://storage.googleapis.com/golang/go1.3.1.linux-amd64.tar.gz 2>&1 | sed -u 's/.* \([0-9]%\)\ \+\([0-9.]\+.\) \(.*\)/\1\n# Downloading at \2\/s, ETA \3/' | zenity --progress --title="Downloading go1.3.1.linux-amd64.tar.gz" --auto-close --width="400" 
# +adding git support (install,config user,generate ssh keys,add $GOPATH/github.com/user/)
# +liteide git support ($HOME/liteide/share/liteide/litebuild/command/go.api)
#

clear

#get last version of go compiler (e.g. go1.3)
v=`echo $(wget -qO- golang.org) | awk '{ if (match($0,/go[1-9]+.[0-9]+./)) print substr($0,RSTART,RLENGTH) }'`

#get host computer arch (e.g. i386|amd64)
if [[ $(uname -i) == "i386" ]]; then a="386"; else a="amd64"; fi

#get kernel name (e.g. linux|freebsd)
k=$(uname -s | tr '[:upper:]' '[:lower:]')

#build compiler name (e.g. go1.3.linux-386.tar.gz)
n=${v}${k}-${a}.tar.gz

echo "Download last compiler $n..."
wget -Nq -P $HOME/Downloads http://golang.org/dl/$n
echo "Unpack..."
tar -xf $HOME/Downloads/$n -C $HOME

#get last version of ide (e.g. X23.2)
v=`echo $(wget -qO- http://sourceforge.net/projects/liteide/files/) | awk '{ if(match($0,/X[0-9]+.[0-9]+/)) print substr($0,RSTART,RLENGTH) }'`

#get host computer LONG_BIT (e.g 32|64)
a=$(getconf LONG_BIT)

#build ide name (e.g. liteidex23.2.linux-32.tar.bz2)
n=liteidex${v:1}.${k}-${a}.tar.bz2

echo "Download last ide $n..."
wget -Nq -P $HOME/Downloads http://sourceforge.net/projects/liteide/files/${v}/$n
echo "Unpack..."
tar -xf $HOME/Downloads/$n -C $HOME

echo "Get Monaco font..."
wget -Nq -P $HOME/.fonts http://usystem.googlecode.com/files/MONACO.TTF

echo "Create \$GOPATH"
GOPATH=$HOME/go-programs
mkdir -p $GOPATH/src

echo "Install git..."
sudo apt-get install git -y > /dev/null
echo "Setup github"
echo -n "Git user  ";read GITUSER
echo -n "Git email ";read EMAIL
echo -n "Git server [ENTER for github.com]"; read GITSERVER
if [ $GITSERVER=="" ]; then GITSERVER="github.com"; fi

#create git configuration
echo -e "[user]
	name = $GITUSER
	email = $EMAIL" > $HOME/.gitconfig

KEY="rsa"
#if [ ! -f $HOME/.ssh/id_$KEY ];
#then
ssh-keygen -qt $KEY -C "$EMAIL"
# -f $HOME/.ssh/id_$KEY
echo "Copy next key to $GITSERVER/settings/ssh and press any key"
echo
cat $HOME/.ssh/id_$KEY.pub
read
#bug workaround https://help.github.com/articles/error-permission-denied-publickey
eval `ssh-agent -s` > /dev/null
#ssh-add ~/.ssh/id_rsa
echo "Checking the key..."
ssh -o "StrictHostKeyChecking no" -T git@$GITSERVER
#create $GITSERVER in $GOPATH
mkdir -p $GOPATH/src/$GITSERVER/$GITUSER

echo "Add git support to liteide..."
echo -e "git clone git@$GITSERVER:$GITUSER/PROJECT.git $GOPATH/src/$GITSERVER/$GITUSER/PROJECT
git add
git commit -m \"-\" -a
git push" >$HOME/liteide/share/liteide/litebuild/command/go.api

#create GOROOT
GOROOT=$HOME/go

echo "Create liteide.ini.mini"
#create directory for liteide.ini.mini
mkdir -p $HOME/.config/liteide
echo -e "[liteenv]
currentenv=linux$a
\n[liteide]
gopath=$GOPATH
\n[liteapp]
geometry=\"@ByteArray(\x1\xd9\xd0\xcb\0\x1\0\0\0\0\0\x41\0\0\0\x18\0\0\x5U\0\0\x2\xff\0\0\0\0\0\0\0,\0\0\x3\x17\0\0\x2\xd3\0\0\0\0\x2\0)\"
state=@ByteArray(\0\0\0\xff\0\0\0\0\xfd\0\0\0\x4\0\0\0\0\0\0\x1&\0\0\x2\x8d\xfc\x2\0\0\0\x2\xfb\0\0\0\f\0\x64\0o\0\x63\0k\0_\0\x31\0\0\0\0\0\xff\xff\xff\xff\0\0\0 \0\xff\xff\xff\xfb\0\0\0\x18\0\x64\0o\0\x63\0k\0_\0\x31\0_\0s\0p\0l\0i\0t\x1\0\0\0\x39\0\0\x2\x8d\0\0\0\xab\0\xff\xff\xff\0\0\0\x1\0\0\x1\x1d\0\0\x2\x8d\xfc\x2\0\0\0\x2\xfb\0\0\0\f\0\x64\0o\0\x63\0k\0_\0\x32\0\0\0\0\x39\0\0\x2\x8d\0\0\0 \0\xff\xff\xff\xfb\0\0\0\x18\0\x64\0o\0\x63\0k\0_\0\x32\0_\0s\0p\0l\0i\0t\0\0\0\0\0\xff\xff\xff\xff\0\0\0 \0\xff\xff\xff\0\0\0\x2\0\0\0\0\0\0\0\0\xfc\x1\0\0\0\x2\xfb\0\0\0\f\0\x64\0o\0\x63\0k\0_\0\x34\0\0\0\0\0\xff\xff\xff\xff\0\0\0^\0\xff\xff\xff\xfb\0\0\0\x18\0\x64\0o\0\x63\0k\0_\0\x34\0_\0s\0p\0l\0i\0t\0\0\0\0\0\xff\xff\xff\xff\0\0\0^\0\xff\xff\xff\0\0\0\x3\0\0\x4\xdb\0\0\0@\xfc\x1\0\0\0\x2\xfb\0\0\0\f\0\x64\0o\0\x63\0k\0_\0\x38\0\0\0\0\x1d\0\0\x4\xdb\0\0\0^\0\xff\xff\xff\xfb\0\0\0\x18\0\x64\0o\0\x63\0k\0_\0\x38\0_\0s\0p\0l\0i\0t\0\0\0\0\0\xff\xff\xff\xff\0\0\0^\0\xff\xff\xff\0\0\x3\xe9\0\0\x2\x8d\0\0\0\x4\0\0\0\x4\0\0\0\b\0\0\0\b\xfc\0\0\0\x3\0\0\0\0\0\0\0\x1\0\0\0\f\0t\0o\0o\0l\0_\0\x31\x2\0\0\0\0\xff\xff\xff\xff\0\0\0\0\0\0\0\0\0\0\0\x1\0\0\0\x1\0\0\0\f\0t\0o\0o\0l\0_\0\x32\x2\0\0\0\0\xff\xff\xff\xff\0\0\0\0\0\0\0\0\0\0\0\x2\0\0\0\x4\0\0\0\f\0t\0o\0o\0l\0_\0\x34\0\0\0\0\0\xff\xff\xff\xff\0\0\0\0\0\0\0\0\0\0\0\x16\0t\0o\0o\0l\0\x62\0\x61\0r\0/\0s\0t\0\x64\0\0\0\0\0\xff\xff\xff\xff\0\0\0\0\0\0\0\0\0\0\0\x16\0t\0o\0o\0l\0\x62\0\x61\0r\0/\0\x65\0n\0v\0\0\0\0\0\xff\xff\xff\xff\0\0\0\0\0\0\0\0\0\0\0\x18\0t\0o\0o\0l\0\x62\0\x61\0r\0/\0t\0\x61\0\x62\0s\x1\0\0\0\0\xff\xff\xff\xff\0\0\0\0\0\0\0\0)
toolState=@ByteArray(\0\0\xff\xe0\0\0\0\0\0\0\0\xe\0\x66\0o\0l\0\x64\0\x65\0r\0s\0\0\0\x1\0\0\0\0\0\x10\0\x65\0v\0\x65\0n\0t\0l\0o\0g\0\0\0\b\x1\0\0\0\0\x18\0s\0\x65\0\x61\0r\0\x63\0h\0r\0\x65\0s\0u\0l\0t\0\0\0\b\x1\0\0\0\0\x16\0\x62\0u\0i\0l\0\x64\0o\0u\0t\0p\0u\0t\0\0\0\b\0\0\0\0\0\xe\0o\0u\0t\0l\0i\0n\0\x65\0\0\0\x2\0\0\0\0\0\x12\0\x63\0l\0\x61\0s\0s\0v\0i\0\x65\0w\0\0\0\x2\0\0\0\0\0\x1a\0W\0\x65\0\x62\0K\0i\0t\0\x42\0r\0o\0w\0s\0\x65\0r\0\0\0\b\0\0\0\0\0\x16\0H\0t\0m\0l\0P\0r\0\x65\0v\0i\0\x65\0w\0\0\0\x2\0\0\0\0\0\x16\0\x64\0\x65\0\x62\0u\0g\0o\0u\0t\0p\0u\0t\0\0\0\b\0\0\0\0\0\x1a\0g\0o\0p\0\x61\0\x63\0k\0\x62\0r\0o\0w\0s\0\x65\0r\0\0\0\x1\x1\0\0\0\0\x18\0g\0o\0\x64\0o\0\x63\0/\0s\0\x65\0\x61\0r\0\x63\0h\0\0\0\x1\x1\0\0\0\0\x12\0g\0o\0\x64\0o\0\x63\0/\0\x61\0p\0i\0\0\0\x1\x1\0\0\0\0\x14\0\x66\0i\0l\0\x65\0s\0y\0s\0t\0\x65\0m\0\0\0\x1\x1\x1)
ShowEditToolbar=false
FolderShowHidenFiles=false
MaxRecentFiles=16
AutoLoadLastSession=false
SplashVisible=false
EditTabsClosable=true
StartupReloadFiles=true
StartupReloadFolders=true
FileWatcherAutoReload=false
\n[golangdoc]
goroot=$GOROOT
\n[FileBrowser]
root=$GOPATH/src
synceditor=false
\n[FileManager]
initpath=$HOME
\n[%General]
Language=en_US
WelcomePageVisible=false
\n[editor]
family=Monaco
fontsize=12" > $HOME/.config/liteide/liteide.ini.mini

#create generic .desktop file on desktop
echo -e "[Desktop Entry]
Version=1.0
Encoding=UTF-8
Name=LiteIDE
Comment=LiteIDE is a simple, open source, cross-platform Go IDE. 
Exec=sh -c 'cp .config/liteide/liteide.ini.mini .config/liteide/liteide.ini; eval \`ssh-agent -s\`; $HOME/liteide/bin/liteide'
Icon=$GOROOT/doc/gopher/gophercolor.png
Type=Application
Categories=Network;" > $HOME/Desktop/liteide.desktop

echo "Create smart launcher"
case $DESKTOP_SESSION in
ubuntu*)
   #create directory for liteide.desktop
   mkdir -p $HOME/.local/share/applications/

   #extend .desktop file with nice options
   echo -e "\nActions=golang;http server;gopath;github;
\n[Desktop Action golang]
Name=golang.org
Exec=firefox golang.org/pkg %U
OnlyShowIn=Unity;
\n[Desktop Action http server]
Name=HTTP server (localhost:8080)
Exec=firefox localhost:8080
OnlyShowIn=Unity;
\n[Desktop Action gopath]
Name=\$GOPATH
Exec=nautilus go-programs/src %U
OnlyShowIn=Unity;
\n[Desktop Action github]
Name=github.com/$GITUSER
Exec=firefox github.com/$GITUSER %U
OnlyShowIn=Unity;" >> $HOME/Desktop/liteide.desktop
   #add .desktop file to dash and integrate with unity
   mv $HOME/Desktop/liteide.desktop $HOME/.local/share/applications
   #get the current launcher favorites list
   b=$(gsettings get com.canonical.Unity.Launcher favorites)
   #skip update if liteide launcher already exists
   if ! [[ $b =~ "liteide.desktop" ]]; then
      #eye candy if is put after firefox or nautilus :)
      if [[ $b =~ "firefox.desktop" ]]; then
         #check if firefox has a launcher and insert liteide after
         b=${b/\'firefox.desktop\'/ \'firefox.desktop\', \'liteide.desktop\'}
      elif [[ $b =~ "nautilus-home.desktop" ]]; then
         #if someone does not use firefox, insert after home folder
         b=${b/\'nautilus-home.desktop\'/ \'nautilus-home.desktop\', \'liteide.desktop\'}
      else
         #otherwise, insert after all favorites
         b=${b/]/, \'liteide.desktop\']}
      fi
      #update the launcher favorites list. in unity changes are shwown immediately.
      gsettings set com.canonical.Unity.Launcher favorites "$b"
      #unity2d need restart to show last changes
      if [[ $DESKTOP_SESSION =~ "2d" ]]; then
         killall unity-2d-shell;
      fi
   fi
   ;;
#other desktop environments can be handled here
*)
   #generic desktop environment have only a desktop shortcut
   chmod +x $HOME/Desktop/liteide.desktop
   ;;
esac

echo "Create HelloWorld program"
mkdir -p $GOPATH/src/HelloWorld
echo -e "package main\n
func main() {
	println(\"Hello World!\")
}" > $GOPATH/src/HelloWorld/main.go

echo "Done."


Install.sh
#!/bin/bash

# golang programming environment installer
# Copyright (C) 2014-2016  geosoft1@gmail.com
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

#B0006
set -e

#if dumb terminal (file browser) run xterm
if [ $TERM == "dumb" ]; then
   xterm -hold -e $0
fi

#-------------------------------------------------------------------------------

clear

echo "Golang Programming Environment Installer"
echo "Copyright (C) 2014-2016  geosoft1@gmail.com"

usage()  {
   echo "Usage: "`basename "$0"`" [options]"
   echo "Options:"
   echo "-c        enable classroom mode"
   echo "-e        install emergency ide version"
   echo "-g        enable git suppport"
   echo "-h        show this help message and exit"
   echo "-s        install system only, no qt libs"
   echo "-u        uninstall"
   echo "-U        uninstall include .gitconfig file and .ssh folder"
   echo "-v        version"
   exit
}

while getopts ":e:cghsuUrv" OPTION; do
case $OPTION in
   c ) CLASSROOM=yes;;
   e ) EMERGENCY=$OPTARG/;;
   g ) GITSUPPORT=yes;;
   h ) usage;;
   s ) QTLIBS=no;;
   U ) rm -rf $HOME/.local/share/data/liteide/
       rm -f  $HOME/.gitconfig
       rm -rf $HOME/.ssh/
       ;&
   u ) rm -rf $HOME/liteide/
       rm -rf $HOME/go/
       rm -rf $HOME/.local/share/applications/liteide.desktop
       rm -rf $HOME/.dlv/
       rm -rf $HOME/.config/gocode/
       #B0020
       rm -f  $HOME/.local/share/data/liteide/*.*
       rm -rf $HOME/.config/liteide/
       rm -rf $HOME/.fonts/MONACO.TTF
       sed --in-place '/export GOROOT=$HOME\/go/d' $HOME/.bashrc
       sed --in-place '/export PATH=$PATH:$GOROOT\/bin/d' $HOME/.bashrc
       sed --in-place '/export GOPATH=$HOME\/go-programs/d' $HOME/.bashrc
       echo "Uninstalled."
       exit;;
   v ) echo ${VERSION=1.0.5.1}; exit;;

   \?) echo "Unknown option: -$OPTARG"; exit;;
   : ) echo "Missing option argument for -$OPTARG"; exit;;
   * ) echo "Unimplemented option: -$OPTARG"; exit;;
   esac
done

#-------------------------------------------------------------------------------

#get host computer arch (e.g. i686|amd64)
#B0002,B0007
case $(uname -m) in
i686 ) a="386";;
   * ) a="amd64";;
esac

#get host computer LONG_BIT (e.g 32|64)
b=$(getconf LONG_BIT)

#get kernel name (e.g. linux|freebsd)
k=$(uname -s | tr '[:upper:]' '[:lower:]')

#get localized user directories
#B0005
test -f ${XDG_CONFIG_HOME:-~/.config}/user-dirs.dirs && source ${XDG_CONFIG_HOME:-~/.config}/user-dirs.dirs

#-------------------------------------------------------------------------------

#get last version of go compiler (e.g. go1.5.1.)
#B0009,B0021
v=`echo $(wget --no-check-certificate -qO- golang.org) | 
awk '{ 
   if (match($0,/go([0-9]+.)+/)) 
      print substr($0,RSTART,RLENGTH) 
}'`

#exit if no network connection otherwise the rest will fail
#B0003 
if [ -z "$v" ]; then
    echo "No network connection"; exit
fi

#build compiler name (e.g. go1.5.1.linux-386.tar.gz)
n=${v}${k}-${a}.tar.gz
#n=$v$k-$a.tar.gz

echo "Download last compiler $n..."
#ERROR: certificate common name `*.googleusercontent.com' doesn't match requested host name `storage.googleapis.com'.
#To connect to storage.googleapis.com insecurely, use `--no-check-certificate'.
#B0004
wget --no-check-certificate -qNP ${XDG_DOWNLOAD_DIR} https://storage.googleapis.com/golang/$n

echo "Unpack..."
tar -xf ${XDG_DOWNLOAD_DIR}/$n -C $HOME

#-------------------------------------------------------------------------------

#determine last version of ide (e.g. X27.2.1) from sourceforge
#B0009,B0011
v=$EMERGENCY
if [ -z "$v" ]; then
v=`echo $(wget -qO- http://sourceforge.net/projects/liteide/files/) | 
awk '{ 
   if(match($0,/X([0-9]+.)+/)) 
      print substr($0,RSTART,RLENGTH) 
}'`
fi

#exit if site is offline otherwise the rest will fail 
if [ -z "$v" ]; then
   echo "website is temporarily in static offline mode."; exit
fi

#build ide name (e.g. liteidex27.2.1.linux-32-qt4-system.tar.bz2)
#B0015,B0018,B0022
n=`echo $(wget -qO- http://sourceforge.net/projects/liteide/files/$v) | 
awk '{ 
   if(match($0, /liteidex([-. '${v:1:-1}' '$k' qt(4|5) '$b'])+tar.bz2/ )) { 
      print substr($0,RSTART,RLENGTH) 
   } 
}'`

#exit if site is offline otherwise the rest will fail 
if [ -z "$n" ]; then
   echo "website is temporarily in static offline mode."; exit
fi

echo "Download last ide $n..."
wget -qNP ${XDG_DOWNLOAD_DIR} http://sourceforge.net/projects/liteide/files/$v$n

echo "Unpack..."
tar -xf ${XDG_DOWNLOAD_DIR}/$n -C $HOME

#-------------------------------------------------------------------------------

if [ -n "$QTLIBS" ]; then
   rm $HOME/liteide/lib/liteide/libQt*.*
fi

echo "Get Monaco font..."
wget -nc -qP $HOME/.fonts http://usystem.googlecode.com/files/MONACO.TTF

echo "Create \$GOPATH"
GOPATH=$HOME/go-programs
mkdir -p $GOPATH/src

#create GOROOT
GOROOT=$HOME/go

#-------------------------------------------------------------------------------

echo "Add git support to liteide..."
if [ -n "$GITSUPPORT" ]; then
   GITSERVER="github.com"
   #install git and curl if not installed
   #B0017
   if ! which git > /dev/null; then
      echo "Install git..."
      sudo apt-get install git curl -y > /dev/null
   fi
   #create git configuration if not exist, otherwise use existent (~/.gitconfig)
   if [ -f $HOME/.gitconfig ]; then
      GITUSER=`awk 'NR==2 {print $3}' $HOME/.gitconfig`
      GITEMAIL=`awk 'NR==3 {print $3}' $HOME/.gitconfig`
   else
      echo "Setup git"
      echo -n "Git user ";read GITUSER
      echo -n "Git email [ENTER for $GITUSER@gmail.com] ";read GITEMAIL
      #try to guess git email
      if [ -z "$GITEMAIL" ]; then GITEMAIL="$GITUSER@gmail.com"; fi
      git config --global user.name "$GITUSER"
      git config --global user.email "$GITEMAIL"
   fi
   #generate ssh keys if not exist, otherwise use existent (https://help.github.com/articles/generating-ssh-keys/)
   KEYTYPE="rsa"
   if [ ! -f $HOME/.ssh/id_$KEYTYPE ]; then
      ssh-keygen -qt $KEYTYPE -C "$GITEMAIL" -f $HOME/.ssh/id_$KEYTYPE
      #add a new deploy key on github with api (https://developer.github.com/v3/)
      echo -n "Password:"; read -s GITPASSWORD
      echo 
      KEY=`cat $HOME/.ssh/id_$KEYTYPE.pub`
      #Key management:
      #curl -s -X GET -u $GITUSER:$GITPASSWORD https://api.github.com/user/keys
      #curl -s -X DELETE -u $GITUSER:$GITPASSWORD https://api.github.com/user/keys/13146480
      #curl -s -X POST -u $GITUSER:$GITPASSWORD https://api.github.com/user/keys -d '{"key":"'"${KEY}"'"}'
      err=`curl -s -X POST -u $GITUSER:$GITPASSWORD https://api.$GITSERVER/user/keys -d '{"key":"'"${KEY}"'"}'| awk '/message/ { gsub(/^[\t ]+|[\",]/,"");print }'`
      if [ "$err" != "" ]; then
         echo -e $err
      fi
   fi
   #bug workaround https://help.github.com/articles/error-permission-denied-publickey
   eval `ssh-agent -s` > /dev/null
   echo "Checking the keys..."
   #workaround: if ssh result is false (Permission denied (publickey).) set -e will stop the script
   #prevent this by changing result code to true and let the script to continue
   ssh -o StrictHostKeyChecking=no -o LogLevel=error -T git@$GITSERVER || true
   #create github.com in $GOPATH
   mkdir -p $GOPATH/src/$GITSERVER/$GITUSER
   #show gopei shell mode :-)
   #wget -q https://raw.githubusercontent.com/geosoft1/tools/master/gopher/gopeicolor.png -O $GOROOT/doc/gopher/gophercolor.png

   #setup desktop action to launcher
   GITACTION="$GITSERVER;"
   GITDESKTOPACTION="[Desktop Action $GITSERVER]
Name=$GITSERVER/$GITUSER
Exec=xdg-open http://$GITSERVER/$GITUSER"
fi

#-------------------------------------------------------------------------------

#get other usefull tools
wget -q https://raw.githubusercontent.com/geosoft1/tools/master/resources/scripts/clone -O $HOME/liteide/bin/clone
wget -q https://raw.githubusercontent.com/geosoft1/tools/master/resources/scripts/repo -O $HOME/liteide/bin/repo
wget -q https://raw.githubusercontent.com/geosoft1/tools/master/resources/scripts/get -O $HOME/liteide/bin/get
chmod +x $HOME/liteide/bin/clone
chmod +x $HOME/liteide/bin/repo
chmod +x $HOME/liteide/bin/get

#build essential git commands list
echo -e "git commit -m \"-\" -a
git push
git pull
git add *
clone
repo
get
present" >$HOME/liteide/share/liteide/litebuild/command/go.api

#-------------------------------------------------------------------------------

#create system environment for ide
#B0012,B0019
echo -e 'GOARCH='$a'\nGOROOT=$HOME/go\nPATH=$PATH:$GOROOT/bin' >> $HOME/liteide/share/liteide/liteenv/system.env

echo "Create liteide.ini.mini"
#create directory for liteide.ini.mini
mkdir -p $HOME/.config/liteide
#get liteide.ini.mini
wget -q https://raw.githubusercontent.com/geosoft1/tools/master/resources/liteide.ini.mini -O $HOME/.config/liteide/liteide.ini
sed -i "s#\$a#$a#g; s#\$GOPATH#$GOPATH#g; s#\$GOROOT#$GOROOT#g; s#\$HOME#$HOME#g" $HOME/.config/liteide/liteide.ini
if [ "$CLASSROOM" == "yes" ]; then
   #B0016
   cp $HOME/.config/liteide/liteide.ini $HOME/.config/liteide/liteide.ini.mini
   #add customizer command
   CUSTOMIZER="cp $HOME/.config/liteide/liteide.ini.mini $HOME/.config/liteide/liteide.ini;"
fi
#sublime theme workaround
echo -e "QAbstractScrollArea {\n\tborder: 0px;\n}\nQTreeView {\n\tborder: 1px solid #cccccc;\n}" > $HOME/liteide/share/liteide/liteapp/qss/sublime.qss

#create generic .desktop file on desktop
echo -e "[Desktop Entry]
Version=1.0
Name=LiteIDE
Comment=LiteIDE is a simple, open source, cross-platform Go IDE. 
Exec=sh -c 'eval \`ssh-agent -s\`;$CUSTOMIZER$HOME/liteide/bin/liteide'
Icon=$GOROOT/doc/gopher/gophercolor.png
Type=Application" > ${XDG_DESKTOP_DIR}/liteide.desktop

echo "Create smart launcher"
case $DESKTOP_SESSION in
ubuntu*)
   #create directory for liteide.desktop
   mkdir -p $HOME/.local/share/applications/
   #B0014
   #extend .desktop file with nice options
   echo -e "Actions=golang;http;gopath;$GITACTION
\n[Desktop Action golang]
Name=golang.org
Exec=xdg-open http://golang.org/pkg
\n[Desktop Action http]
Name=HTTP server (localhost:8080)
Exec=xdg-open http://localhost:8080
\n[Desktop Action gopath]
Name=GOPATH
Exec=xdg-open go-programs/src
\n$GITDESKTOPACTION" >> ${XDG_DESKTOP_DIR}/liteide.desktop

   #add .desktop file to dash and integrate with unity
   mv ${XDG_DESKTOP_DIR}/liteide.desktop $HOME/.local/share/applications
   #get the current launcher favorites list
   b=$(gsettings get com.canonical.Unity.Launcher favorites)
   #skip update if liteide launcher already exists
   if ! [[ $b =~ "liteide.desktop" ]]; then
      b=${b/]/, \'liteide.desktop\']}
      #B0001
      sleep 1
      #update the launcher favorites list. in unity changes are shwown immediately.
      gsettings set com.canonical.Unity.Launcher favorites "$b"
   fi;;
#other desktop environments can be handled here
*)
   #generic desktop environment have only a desktop shortcut
   #B0008
   chmod +x ${XDG_DESKTOP_DIR}/liteide.desktop;;
esac

echo "Create some useful templates"
TEMPL=$HOME/liteide/share/liteide/liteapp/template

#create simple template
sed -i '1i gosimple' $TEMPL/project.sub
mkdir -p $TEMPL/gosimple

echo -e "// \$ROOT\$ project
package main\n
func main() {
}" > $TEMPL/gosimple/main.go

echo -e "[SETUP]
NAME = \"Go1 Simple Project\"
AUTHOR = geosoft1
INFO = new go1 simple project
TYPE = gopath
FILES = main.go
OPEN = main.go
SCHEME=folder" > $TEMPL/gosimple/setup.inf

#create gpl template
sed -i '2i gogpl' $TEMPL/project.sub
mkdir -p $TEMPL/gogpl

year=`date +"%Y"`
author=$GITUSER
email=$GITEMAIL
if [ "$author" == "" ]; then
   author="<name of the author>"
fi
if [ "$email" == "" ]; then
   email="<email>"
fi
echo -e "// \$ROOT\$ project
// Copyright (C) $year  $author  $email
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
\npackage main\n
func main() {
}" > $TEMPL/gogpl/main.go

echo -e "[SETUP]
NAME = \"Go1 GPL Project\"
FILES = main.go LICENSE CONTRIBUTORS README.md
AUTHOR = geosoft1
INFO = new go1 gpl project
TYPE = gopath
OPEN = main.go
SCHEME=folder" > $TEMPL/gogpl/setup.inf

wget -Nq -O $TEMPL/gogpl/LICENSE http://www.gnu.org/licenses/gpl.txt
#touch $TEMPL/gogpl/CONTRIBUTORS
echo $GITEMAIL > $TEMPL/gogpl/CONTRIBUTORS
echo -e "\$ROOT\$\n====" > $TEMPL/gogpl/README.md

echo "Create HelloWorld program"
mkdir -p $GOPATH/src/HelloWorld
echo -e "package main\n
func main() {
\tprintln(\"Hello World!\")
}" > $GOPATH/src/HelloWorld/main.go

#add GOPATH,GOROOT to PATH but in $HOME/.bashrc to avoid root rights
#B0013
grep -q 'export GOROOT=$HOME\/go' $HOME/.bashrc || sed -i '$ a\export GOROOT=$HOME\/go' $HOME/.bashrc
grep -q 'export PATH=$PATH:$GOROOT\/bin' $HOME/.bashrc || sed -i '$ a\export PATH=$PATH:$GOROOT\/bin' $HOME/.bashrc
grep -q 'export GOPATH=$HOME\/go-programs' $HOME/.bashrc || sed -i '$ a\export GOPATH=$HOME\/go-programs' $HOME/.bashrc

echo "Done."
